File: 19-RemoveNthNodeFromEnd.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        left_ptr = dummy
        right_ptr = head
        
        while n>0 and right_ptr:
            right_ptr = right_ptr.next
            n-=1
            
        while right_ptr:
            left_ptr = left_ptr.next
            right_ptr = right_ptr.next
            
        left_ptr.next = left_ptr.next.next
        return dummy.next
    
    
    #space - o(1)
    # time - o(n)

File: 19-removeNthNodeFromEnd.java
//Approach-1: Using the two pointers approach
//Overall Time Complexity: O(N)
//Overall Space Complexity: O(1)
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null || head.next == null) return head;
        
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        
        ListNode slow = dummy;
        ListNode fast = dummy;
        
        for(int i = 0; i<=n; i++){
            fast = fast.next;
        }
        
        while(fast != null){
            slow = slow.next;
            fast = fast.next;
        }
        
        slow.next = slow.next.next;
        
        return dummy.next;
    }
}